gennet-module/Dockerfile

```diff
@@ -8,7 +8,7 @@ RUN python -m venv /opt/venv
 ENV PATH="/opt/venv/bin:$PATH"
 
 # Perform the installs
-COPY requirements.txt .
+COPY gennet-module/requirements.txt .
```

Comment:
I don't agree with this change. A project (or submodule in this case) is thought to be and work standalone. For me it makes no sense that it nees to call itself from "outside". It should matter if the folder is `gennet-module` or whatever. If for whatever reason you want to rename the folder, you need to rename the dockerfile also.

Reply:
I have been thinking about this after our meeting today.

I still think the same. Ideally, gennet is a submodule that was thought to be used either inside wakurtosis or outside. Independently about this, it makes no sense to use Docker to search for external resources. 

In this specific case, the line
`COPY gennet-module/requirements.txt .` should be `COPY requirements.txt .`, because the workspace that it is in, should be where the Dockerfile is. We can argue if this could be in an inner level, but never in an upper level.

So basically the question is where should we locate `traits`. And if they will not be used anywhere else, I think those traits should be inside gennet, as they are specifically for the toml generation that gennet does.

Reply:
Your reasoning is indeed sound: 'traits' *ideally* should be under 'gennet-module'.

But putting it under 'config' will allow us to modify 'traits' on the fly. For instance, if I want to  turn on and off storage trait between two runs --- as both 'config.json' and 'traits' are under the same mount point --- the updates are picked up into the container WITHOUT a docker rebuild. But if we move 'traits' exclusively to 'gennet-module',  we *will* have to rebuild the gennet module every time we so much as touch the traits.

So, putting 'traits' under 'config' does offend the *strict*  layering a bit, but the flexibility and efficiency this set up brings makes it worth it. A middle ground is to add 'traits' under  'gennet-modul'e and add a symbolic under to 'config', or vice versa; whichever works.

Reply:
Hm, I am missing something. Can you put me an example? Because between two runs, you just need to modify `config.json` and delete the `store` trait in `node_type_distribution`. You do not need to rebuild anything, and traits would still be inside gennet.

The idea of wakurtosis is that after you have everything set up (images that you need, create the config file and so on) you just have to modify `config.json` between runs. I don't really see where traits is involved there, and also, it is not "mounted" (in terms of Docker mount) anywhere afaik...

Could you elaborate?

Reply:
Sure. Consider the `store` trait. Let's say we want to study absence of store protocol against a store protocol with a specific `store-message-retention-policy`. This will require us to set different values for the this variable in the `store.toml` file under `traits`. If we put `traits` inside `gennet-module`, any such modification will require a `docker build` for gennet to pick up the  modified tomls; hope that part is clear. But if we put `traits` inside 'config`, gennet will pick up the modified the tomls via the  volume mount aka `--volume/-v` at no cost at all.

Essentially, the `config` will "overlay" the default config via --volume ; and gennet will correctly  pick up the modified tomls from the overlay, at *run time*. In case of no such volume mount, gennet will fallback to the set of traits it is originally packaged with.

Reply:
Oh but modifying trait files shouldn't be an option imo. Trait files should be a template for activating that specific trait. For example, `discv5` is `False` as default, so discv5 trait should have the information to setting it to `True`.

If we want to set specific values like specifying a port, dns domain name and so on, this information should be in `config.json`, under a new key. Toml files should act as template so we do not have to have everything hardcoded in gennet. Possibly there are other options similar to this, but a Dockerfile being aware of its upper level should not be a thing.

Reply:
The traits are not designed to be static. They are dynamic and are modifiable: that's the reason why they are encoded as files under `traits` instead of hard-coded strings in gennet. The traits will eventually have a set of default values, but they are just that: defaults that are modifiable.

Also, remember, gennet allows us to introduce new user-defind traits on the fly. That *will* require modifying traits directory and a rebuild.

We agreed on config.json is for general/gennet/wsl sim configs, tomls for node configs: let's not mix them now.

Reply:
No worries, gennet permits 4 input pathways: defaults, commandlines, config.json and vm mounts; and any combination there off! So it *is* hard to wrap the head around.

You are almost there. The thing is, with traits introduced, gennet *will* require the `traits` inside the container to be an independent package. `config.json` is optional for gennet, but `traits` is not. Gennet can work without a vm mount or even a config.json.

Reply:
I see two options here:

1. If you want gennet to work without a vm mount (which I don't understand why, since its easier for reading files or writing to the host), then we copy traits inside gennet module, and if you want to modify those traits, then you rebuild. Which I don't like it at all.

2. If traits is mandatory to gennet, then we mount it as we are doing rigth now. We have the option to use a `config.json` or not, we can use command line argument or not, but if traits are necessary, then the user just have to mount them inside gennet and that's it. They are modifyable without rebuilding, the process of docker building just need to be aware of what is inside gennet folder, and everything works alrighty. 

And now that I think about... This is what you are already doing but without noticing.
This dockerfile will copy traits inside `/config/traits` in the container. Then, if we do a `docker run -v ${dir}/config/:/config gennet`, we are mounting the `config` host folder on top of the `config` folder that is already in the container, thus this information will be deleted.

My point is, it makes no sense to search for external resources when building gennet. If we want them outside because they can be modifyable, and we do not want to modify them inside `gennet.py`, then the correct approach is to mount them, imo.

Reply:
as we discussed, 1) is not an option. 2)  is what I already do; it was a conscious decision to not to include traits as symlinks don't work in docker.

I did not want to change the build context myself, but this is on of the clean compromise I could come up with; other options, like  including  hard-links to individual trait tomls, write a separate build script for gennet, etc are clunky.

It is best to include known, static dependencies into the docker; and gennet must have its own traits.

Reply:
Ok, let me elaborate extensively because there might be a change that either I am not understanding you, you are not understanding me.

With your current changes, in order to build gennet submodule, the dockerfile is aware of what exists on it's upper levels, since you are doing `COPY config/traits .` and `COPY config/config.json .`, for example.

This is not needed. In gennet container, we expect everything is under `/config`. This will be done mounting a volume as we are doing in `run.sh`, with the following line:

`docker run --name gennet-container -v ${dir}/config/:/config gennet --config-file /config/${wakurtosis_config_file} --output-dir /config/topology_generated`

Here, we are mounting our `config folder` from wakurtosis in `/config` inside the container. This is why also we can access to the topology generated once it is created, since `/config` is a mounted volume, and everything that is created there will be also visible in the host.

So, if we want to add traits, and make them modifiable at a file level as you want, we cannot have them inside `gennet` as we would need to rebuild the image every time we make a change as you are saying. What is the option then? Well, having them in `config`, as we have right now. The thing is that we *already* have access to traits in gennet, since the entire `config` folder is mounted inside the container.

Inside the container, you just need to access to `/config/traits/your_trait.toml`. There is no need to do  `COPY config/traits .`. There is no need to expose the upper level of gennet in the Dockerfile. The image will be built only whith the content of `gennet-module`. And then, the user:
- Can use a `config.json` or not.
- Can give command line arguments or not (as we have default ones).
- **Must** provide a mountpoint. Firstly, because it is easier to collect the results generated from the host. And secondly, and most importantly, The user will need to provide traits. If for some reason you want to also have default-non-modifiable traits, then we can just copy those files into `gennet-module`, and have the folder in the same location. So the user does not have to provide traits. But if the user wants to modify them, well, then he/she will need to mount the folder, and with this, the folder inside the container will be ovewritten with the new traits that the user provided. This is if you want the option to run gennet withouth making the user to do traits. If the user will have to have traits yes or yes, then we assume we will have that folder mounted as a volume, and that's it.

With this:
Dockerfile in gennet is aware only about what it is inside `gennet-module`.
Gennet can still be used as stand-alone container.
User can provide or not config/arguments/traits for gennet, or not.

Reply:
Again, this exactly what the current master already does. This PR just removes the **mandatory** mount (with a very specific subdir structure `/config/traits`). For instance, a simple {"nwaku":100} must work; but  currently, it won't, not unless we do a volume mount. After this PR, it will work; rest remains the same.

On collecting results via mounts (by specifying a  vm mounted subdir  in --output-dir; gennet does NOT write to `config` dir at all otherwise)  is not sound idea as the  final `topology_generated` file in the host will be owned by `root`! This is quite undesirable; please check issue #98.

Could we  leave these deeply subjective stylistics off (for now)? Esp. when we don't have a working metrics collector nor a robust  master. This PR is standing in the way of my metrics PR.

Reply:
This is the last comment I will do, as it is getting again a little bit out of hand.

Of course `/config/traits` is a specific directory inside the container. Because whats inside the container is agnostic to the user. Take as an example cAdvisor. Don't you have to mount several volumens in the paths they tell you to make cAdvisor work? And if for some reason you want to make this directory modifyable, ok, then you let the user mount it anywhere he wants in the container, and put this location as an argument. But this is pointless imo.

I don't understand why we want to get rid of mounting a volume if it is an utility that Docker provide us, actually for situations like this, to change it to a build structure that is not clean and will assume stuff like folder names, file locations and so on without the user being able to modify it.

And regarding #98 , I also don't see the problem. If those files as root are a problem and you want to avoid this, you can use the flag `--user` when you do a docker run. Something like `docker run --user 1000:1000`, or getting dinamically the user id that is running `run.sh` and put it in `docker run` should make those files being created by that user instead of root. So you have it in the host and you don't need to do a `docker cp`.

And as a last point, this is not subjective. If you want the user to build the image you cannot expect certain folder names or stuff like that outside of the context. Because as you said, we wanted this to also be used standalone. And we are killing that with this.

As I mentioned in other PR, I am taking also my time in commenting PRs and requesting changes in order to have a better and more maintainable project. Because if we don't care about this, and going to the extreme case, we could just trhow all our code into a single file, forget about tests, and just go ahead with it as long as it works. I think we should help each other to build a better project, trying always to go for the best quality possible. And this is, imo, mandatory from the most important parts, to the subtle ones, like this one. 

I will approve the PR, feel free to change it, or merge it. Sorry about all the time spent here.